### PR TITLE
fix issue 7804 - Allow __traits(getMember) as a BasicType

### PR DESCRIPTION
**Pull request by ghost**
This the continuation of https://github.com/dlang/dmd/pull/8031.

### CODE REVIEW

File: src/dmd/dsymbolsem.d
Comment (BBasile) [inline src/dmd/dsymbolsem.d]: I don't think this can happen. can  `__trait(getMember) ` return something that's not a type and that's not something that can aliased ?

File: test/fail_compilation/e7804_1.d
Comment (BBasile) [inline test/fail_compilation/e7804_1.d]: the error message should not say "aliased" since there's no alias here.

**Comment by thewilsonator** [comment]
```
Dscanner dmd/mtype.d(5046:25)[warn]: Useless final attribute, functions declared within a final class are never virtual
```

File: src/dmd/mtype.d
Comment (thewilsonator) [inline src/dmd/mtype.d]: This field is not in the C++ header.  It seems unused, is it needed?

Comment (BBasile) [reply src/dmd/mtype.d]: yes i should use it in the alias sem instead of using the loc from the alias.

**Review by thewilsonator — APPROVED** [review]

**Comment by BBasile** [comment]
I'm afraid that this one has been merged a bit too eagerly so you can still review and will push additional correction if needed. For example i just addressed one of the problem pointed out in the old PR just a few hours ago (statement VS declaration).

**Comment by thewilsonator** [comment]
Oops sorry.

**Comment by BBasile** [comment]
In theory i think that you should have used the `auto-merge` label to trigger the bot. Then it wouldn't have worked without the approval of a code owner (reviewers with the locker symbol at the top right corner).

**Comment by thewilsonator** [comment]
mir-algothithm is currently broken though. auto-merge wouldn't do anything for a long time.

**Comment by Geod24** [comment]
Yeah that one definitely needed more review. I for one have comments about the changelog wording and some tests to be added with regards to other traits that the one intended to work.

I understand the desire to be quick but there needs to be a minimal period to allow people to review things, according to the importance of the PR.

**Comment by thewilsonator** [comment]
Oh for some reason I though the one it was replacing was approved, it wasn't. I've been looking at too many PRs, need to take a break.

**Comment by BBasile** [comment]
@Geod24, can you make suggestions for the changelog entry ?

About the other traits, `getMember` is checked during parsing. Later what happens during semantic is then that possible `getMember` errors are not specific to the language enhancement, but are normal traits errors, alias errors or type errors.

I can however add these errors (with the changelog modification PR):

```d
module more;

__traits(getMember, more, "object") not_a_Type; // the trait suceeds but the variable decl not
__traits(getMember, more, "NotExisting") not_existing; // trait error
alias nothing = __traits(getMember, more, "NotExisting"); // trait error          
```

File: changelog/alias_from_trait.dd
Comment (Geod24) [inline changelog/alias_from_trait.dd]: I would reword this part. It's too specific to the compiler implementation.

File: changelog/alias_from_trait.dd
Comment (Geod24) [inline changelog/alias_from_trait.dd]: I'm not a native English speaker but I think the "of" is superfluous here.

File: changelog/alias_from_trait.dd
Comment (Geod24) [inline changelog/alias_from_trait.dd]: I just don't understand this whole paragraph. If you `getMember` on a type, you get a type.
But if you `getMember` on something else (a symbol), you should have an alias to that symbol, no?

File: src/dmd/astbase.d
Comment (Geod24) [inline src/dmd/astbase.d]: DMD style says `const ref` everywhere for `Loc`

File: src/dmd/parse.d
Comment (Geod24) [inline src/dmd/parse.d]: 1) That should not be a parsing error.
2) What about `getOverloads`, `getVirtualMethods`, `getVirtualFunctions`, `getUnittests` and `parent` ?

File: test/runnable/e7804.d
Comment (Geod24) [inline test/runnable/e7804.d]: Why are the tests run in the `static this` instead of the main ??

File: test/runnable/e7804.d
Comment (Geod24) [inline test/runnable/e7804.d]: I think more than `isSame`, setting the variable and calling the functions would be a good test.

File: changelog/alias_from_trait.dd
Comment (BBasile) [reply changelog/alias_from_trait.dd]: No it's not related to the implementation. This describes the language change.

File: changelog/alias_from_trait.dd
Comment (BBasile) [reply changelog/alias_from_trait.dd]: ok, i've entirely removed the last paragraph.

**Pull request merged**
